src/transformers/modeling_bert.py

```diff
@@ -957,7 +957,29 @@ def forward(
             ltr_lm_loss = loss_fct(prediction_scores.view(-1, self.config.vocab_size), lm_labels.view(-1))
             outputs = (ltr_lm_loss,) + outputs
 
-        return outputs  # (masked_lm_loss), (ltr_lm_loss), prediction_scores, (hidden_states), (attentions)
+        return outputs  # (ltr_lm_loss), (masked_lm_loss), prediction_scores, (hidden_states), (attentions)
+
+    def prepare_inputs_for_generation(self, input_ids, attention_mask=None, **model_kwargs):
+        # Add dummy token at the end (no attention on this one)
+
+        assert self.config.pad_token_id is not None, "The PAD token should be defined for generation"
+
+        input_shape = input_ids.shape
+        effective_batch_size = input_shape[0]
+
+        dummy_token = torch.full(
+            (effective_batch_size, 1), self.config.pad_token_id, dtype=torch.long, device=input_ids.device
+        )
+        input_ids = torch.cat([input_ids, dummy_token], dim=1)
+
```

Comment:
The `attention_mask` logic is quite important. The `prepare_inputs_for_generation()` is set up for two cases:
1) Bert is used as a stand-alone decoder. In this case the "correct" `attention_mask` is given by the `generate()` function and will be extended by one
2) Bert is used as a encoder-decoder model (Bert-2-Bert). In this case this function is called for the decoder Bert only and for which we have to extend the `decoder_attention_mask`, which is **not** given (yet) by the `generate()` function. Therefore we have to create it here.